### ISSUE TITLE
Narrow range and `cICP`

### ISSUE THREAD

**Issue by svgeesus** [w3c/png#312]
See [later discussion on narrow range and full range](https://github.com/w3c/png/issues/312#issuecomment-2325281113)

===========

This can be tested in WPT by converting full-range sRGB values to one of the common encodings (such as 2100 PQ, 2100 HLG, Display P3, linear sRGB) and making a PNG with those values and the appropriate `cICP` chunk. A pass displays the same colors as an sRGB reference image.

For non-browser support, detecting and correctly decoding the chunk demonstrates support.

**Comment by svgeesus** [comment]
TweakPNG may add support for `cICP`:

 - https://github.com/jsummers/tweakpng/issues/11#issuecomment-1537449629

**Comment by svgeesus** [comment]
Chromium has `cICP` support for PNG (and also supports CICP in AVIF):

 - [Implement cICP chunk for PNG as per version 3 of the spec](https://chromium-review.googlesource.com/c/chromium/src/+/3705739)

**Comment by ProgramMax** [comment]
What you described is nearly exactly what is already being tested here:
https://wpt.fyi/results/png/cicp-chunk.html?label=master&label=experimental&aligned

There is a `cICP` chunk which marks the image as Display P3.
The image is painted into a canvas and its color value read and compared to what it would be when converted to sRGB.

Does that mean we can close this?

**Comment by svgeesus** [comment]
We could, although I would prefer to see more than one test; and I want to track non-browser implementations like TweakPNG too. So I prefer to keep it open.

**Comment by kmilos** [comment]
FWIW, [darktable ](https://www.darktable.org/) implemented support for cICP chunk some time ago (though only a subset of possible color spaces are actually [built in](https://github.com/darktable-org/darktable/blob/772ec1efb424c1ab08e6f1f3b8a349568d1af09c/src/common/colorspaces.h#L71-L95) and recognized).

**Comment by svgeesus** [comment]
Thanks for the heads-up! Yes I see it in the [darktable 4.2.0 announcement](https://www.darktable.org/2022/12/darktable-4.2.0-released/):

> Attempt to obtain the color-space for PNG files from the cICP chunk. This was added in a recent revision of the PNG spec, so we take advantage of it, if present.

**Comment by svgeesus** [comment]
- [libspng issue for adding `cICP` support](https://github.com/randy408/libspng/issues/218)

**Comment by svgeesus** [comment]
I [raised a Bugzilla issue on Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1862245) for `cICP` support.

**Comment by svgeesus** [comment]
[exiv2 feature request to support `iCCP`](https://github.com/Exiv2/exiv2/issues/2295)

**Comment by svgeesus** [comment]
Mozilla [Request for standards position](https://github.com/mozilla/standards-positions/issues/922) for `cICP`

**Comment by svgeesus** [comment]
The [ImageSharp](https://github.com/SixLabors/ImageSharp) image library for .NET has [added PNG cICP support](https://github.com/SixLabors/ImageSharp/pull/2592/files)

**Comment by svgeesus** [comment]
At the [8 Jan 2024 PNG WG meeting](https://lists.w3.org/Archives/Public/public-png/2024Jan/0013.html) it was concluded that `cICP` has now met the W3C minimum support criteria (two independent implementations) so removing that tag. 

Leaving the issue open to collect latest implementation status, for the implementation report.

**Comment by svgeesus** [comment]
I have a PR for a test that combines APNG and `cICP` to check that all frames inherit the colorspace information, not just the static image.

  https://github.com/web-platform-tests/wpt/pull/43980

**Comment by svgeesus** [comment]
> I have a PR for a test that combines APNG and cICP to check that all frames inherit the colorspace information, not just the static image.

PR is now merged, as hoped [Chrome (and Edge) pass while Firefox and Safari do not](https://wpt.fyi/results/png/apng/fDAT-inherits-cICP.html?label=master&label=experimental&aligned)

**Comment by svgeesus** [comment]
From the [documentation}(github.com/shssoichiro/oxipng/blob/master/MANUAL.txt) for [oxipng](https://github.com/shssoichiro/oxipng), an optimizer written in rust:

> --strip `<mode>`
>          Strip metadata chunks, where `<mode>` is one of:
>         
>          safe    =>  Strip all non-critical chunks, except for the following:
>                         cICP, iCCP, sRGB, pHYs, acTL, fcTL, fdAT
>         all      =>  Strip all non-critical chunks
>          <list>  =>  Strip chunks in the comma-separated list, e.g. 'bKGD,cHRM'
> [source](https://github.com/shssoichiro/oxipng/blob/1b500da585efea50dff33a30694c0923c85f8eb8/MANUAL.txt#L57)

So clearly, it supports `cICP` and knows that stripping it will affect the rendered image.

**Comment by svgeesus** [comment]
CSS Color 4, in the section [Color Space of Tagged Images](https://drafts.csswg.org/css-color-4/#tagged-images), now includes the test results for the PNG cICP tests.

**Comment by svgeesus** [comment]
I added a WPT PR to [test that `cICP` wins over `iCCP`](https://github.com/web-platform-tests/wpt/pull/44644)

**Comment by svgeesus** [comment]
> I added a WPT PR to [test that `cICP` wins over `iCCP`](https://github.com/web-platform-tests/wpt/pull/44644)

This is now merged. Chrome 123 on Linux 20, and Edge 123 on Win 10 both pass. Safari TP 188 on macOS 13.6 is reported as failing, because it needs macOS 14 for CICP support. Firefox 125 on Linux 20 fails because they don't implement `cICP` yet, see [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1862245)

**Comment by svgeesus** [comment]
libavif has an open issue

 - [preserving CICP information when converting AVIF to PNG](https://github.com/AOMediaCodec/libavif/issues/1421)

**Comment by wantehchang** [comment]
@svgeesus Chris: The libavif issue https://github.com/AOMediaCodec/libavif/issues/1421 that you mentioned in https://github.com/w3c/PNG-spec/issues/312#issuecomment-1979051651 covers conversion of AVIF to both PNG and JPEG. The PNG part of that libavif issue is already fixed. So for PNG that issue should be considered closed.

**Comment by svgeesus** [comment]
[Affinity 2.4 supports `cICP` in PNG](https://forum.affinity.serif.com/index.php?/topic/196692-32-bit-hdr-png-support-added/) including narrow and full range.

**Comment by digitaltvguy** [comment]
So should PixelMator/Photomator.  They enabled the MacOS API support in one day after requesting it :-)

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
From: Chris Lilley ***@***.***>
Sent: Wednesday, March 27, 2024 1:12:55 AM
To: w3c/PNG-spec ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [EXTERNAL] Re: [w3c/PNG-spec] Testing `cICP` (Issue #312)


Affinity 2.4 supports cICP in PNG<https://urldefense.com/v3/__https://forum.affinity.serif.com/index.php?*topic*196692-32-bit-hdr-png-support-added*__;Ly8v!!PIZeeW5wscynRQ!otyqRX3QaRGte8sihzbhxkuJW6BicTwlua1KKGSD8H-u4q9pj5bILl1NkGLKGTJKuXdPObVo2yyOiGCkVmPi7qNkPg$> including narrow and full range.

**Comment by fintelia** [comment]
I tested the Video Full Range Flag in both Chrome and Safari. Neither one seems to respect it: pixel values were the same regardless  of whether it was set to zero or one

(Firefox was the same... but that's expected since it doesn't have cICP support yet.)

**Comment by svgeesus** [comment]
I see [several Chrome CICP bugs](https://issues.chromium.org/issues?q=cicp) one of which relates to [Full-Range AVIF](https://issues.chromium.org/issues/354454801). Probably best to open a separate issue for PNG Full/Narrow range CICP.

**Comment by digitaltvguy** [comment]
That’s been true of most applications (even for video wrapped movies).

From: Jonathan Behrens ***@***.***>
Date: Monday, September 2, 2024 at 4:36 PM
To: w3c/png ***@***.***>
Cc: Seeger, Chris (NBCUniversal) ***@***.***>, Comment ***@***.***>
Subject: [EXTERNAL] Re: [w3c/png] Testing `cICP` (Issue #312)

I tested the Video Full Range Flag in both Chrome and Safari. Neither one seems to respect it: pixel values were the same regardless of whether it was set to zero or one

(Firefox was the same... but that's expected since it doesn't have cICP support yet.)

**Comment by ProgramMax** [comment]
This feels like a tricky one.
For typical computer use cases today, all images will be full range. In fact, I think even an (non-broadcast) expert would need to work very hard to find even 1 real world narrow range image. As a result, I can understand not wanting to spend engineering time solving a non-problem.
[Chrome considered it explicitly and closed as won't-fix because there was no interest](https://issues.chromium.org/issues/40229816).

However, that then forces the question "Then should it be part of the spec?"

However, for legacy broadcast, it does seem important. That's why it is part of H.273 in the first place.
For broadcast experts, I could see PNG test images that use this flag being very useful. Any software along the pipeline which reads CICP data will want it.

To me, the tradeoff is "Almost all users don't need this, so we should drop it..." vs. "...but then we're not using real CICP / H.273."

Perhaps we could still reference H.273 for all CICP values and just drop that field.
(To that end, perhaps we can also drop the one that says the color model is RGB, since we likely will always have a standard RGB PNG available even if we later add extra channels.)
But I feel like the cost of 2 bytes which are unenforced is worth it to keep H.273 in tact. What if later H.273 changes from 4 bytes to 2 bit version flag, 6bits, 12bits, 4 bits, 8 bits? We'll have wished we kept all 32 bits.

So maybe the key is for us to make it a note and reference H.273 but to not require strict enforcement on the values?

**Comment by fintelia** [comment]
The difference between a cICP chunk taking 16 bytes versus 14 bytes (once counting the chunk header) doesn't really seem like enough justification to diverge from the base specification.

I'd say that the spec should only require that decoders support full-range images, given that that's what major implementations are planning on doing. Leaving the text as-is would likely create pointless engineering work for implementers who don't realize quite how rare narrow-range images are. Though I'm not quite sure how to word it in the spec so that web browsers which ignore the flag as well as legacy broadcast folks' software can both be considered conformant.

**Comment by svgeesus** [comment]
If we say that narrow-range images are poorly supported and should not be created, we effectively require creation/conversion tools to convert narrow-range images to full range, which is lossy due to quantization error.

**Comment by svgeesus** [comment]
The [definition of reference image](https://w3c.github.io/png/#3referenceImage) is a full-range image (although it does not explicitly use that term, and probably should): it clearly maps 0, not 16, to black for example.

**Comment by svgeesus** [comment]
The [cICP section of the Implementation Report](https://w3c.github.io/png/Implementation_Report_3e/#cicp) sumarizes overall `cICP` status well, so renaming this issue to capture the shift to narrow/full range discussion.

**Comment by svgeesus** [comment]
Firefox [just added support for `cICP` in PNG](https://hg.mozilla.org/mozilla-central/rev/8599fc020cd6) and the code looks for narrow range, and gives a warning that it is not supported.

**Comment by tnikkel** [comment]
> Firefox [just added support for `cICP` in PNG](https://hg.mozilla.org/mozilla-central/rev/8599fc020cd6) and the code looks for narrow range, and gives a warning that it is not supported.

I made that decision based largely on the discussion here that such images were basically non-existent and Chrome does not support them. It didn't seem like a good use of resources at this time to add support for something that doesn't exist. If those type of images become more common we can revisit this.

**Comment by svgeesus** [comment]
WPT tests `cicp-chunk`, `cICP-wins` and `fDAT-inherits-cICP` are now [all passed in Firefox nightly 136](https://wpt.fyi/results/png?label=experimental&label=master&aligned)

**Comment by svgeesus** [comment]
> If we say that narrow-range images are poorly supported and should not be created, we effectively require creation/conversion tools to convert narrow-range images to full range, which is lossy due to quantization error.

Although, I believe most narrow-range data sources are YCbCr video which requires conversion to RGB anyway, to store as PNG

**Comment by jbowler** [comment]
@svgeesus, you commented:

>If we say that narrow-range images are poorly supported and should not be created, we effectively require creation/conversion tools to convert narrow-range images to full range, which is lossy due to quantization error.

Yeah.  But if implementations do not support the "poorly supported" the poorly supported is not supported.

Subsequent discussions in this thread suggest that no one who actually implements support will do so.  Think about that and, please, respond.  If nothing else it concerns me bigly.

**Comment by ProgramMax** [comment]
Implementations do support narrow-range. It is well supported. There are [standard test files, even](https://github.com/digitaltvguy/PNG-TIFF-MOV-MXF-cICP-mDCV-cLLI-Conformance-Files/blob/main/PNG/PNG%20-%20with%203rd%20Edition%20cICP%20mDCV%20cLLI%20Chunks/PNG-SDR-BT.709-ColorBars-Tent-Valley-Grayscale-16bit-cICP-mDCV-100nit-.01nit-Narrow_Range.png).

It isn't used in the computer space. But it is (well, was) heavily used in the broadcast space.

I don't think we need to provide extra tooling in the computer space (where this isn't used). And I think it is fine for the computer space to add it if/when there is demand in the computer space (which is probably never).

**Comment by svgeesus** [comment]
I'm reading (for example [here](https://forum.affinity.serif.com/index.php?/topic/196692-32-bit-hdr-png-support-added/)) that Final Cut Pro and DaVinci Resolve support 32bit HDR PNG with `cICP` including narrow and full range.However the product pages

 - [Final Cut Pro](https://www.apple.com/final-cut-pro/)
 - [DaVinci Resolve](https://www.blackmagicdesign.com/products/davinciresolve/?pStoreID=massmutual)

are breezy brochure-like overviews and don't confirm this in detail.

**Comment by svgeesus** [comment]
PixelMator Pro now has a [list of supported formats](https://support.pixelmator.com/faq-pixelmator-pro/advanced-workflows/working-with-hdr-content-in-pixelmator-pro) including HDR PNG both for import and for export.

**Comment by jbowler** [comment]
> I don't think we need to provide extra tooling in the computer space (where this isn't used).

Note that @svgeesus 's [link](https://forum.affinity.serif.com/index.php?/topic/196692-32-bit-hdr-png-support-added/) from the "closed as completed" comment calls this statement into question.  In particular the comment from Graphite Addict of Lincoln on February 29 show that 16-bit PNG exports from Affinity Designer which used to export without HDR (and no cICP) now export with cICP and HDR after the addition of support to "all" Affinity apps without user changes (apparently).  See the subsequent comments from Affinity employees who responded.

Then the user "Frozen Death Knight" comments:

>I tried exporting an image with HDR. However, what I noticed was that viewing the exported image outside of Affinity made the colours more dull and less bright when using the built in Photo viewer on Windows 10. When importing the new .png-file back into Affinity the colours are correct again with the HDR preview on.

That indicates that there is no cICP support yet in GDI+; Windows GDI+ has native PNG decoding but I would expect Microsoft to wait on the REC.  I.e. to implement it in-house but not do a release until the REC is out.

In both cases the user does not have any expertise and, in some sense, didn't know what they were doing.  Since Affinity does support both narrow and full range there will be narrow range images in the "computer space".  See FDK's screenshot of the UI; the "Full Range" flag is in "Advanced" and it is on but given that most people won't have the faintest idea what it does I imagine it will get switched off a lot.

Graphite Addict also found that the BT.709 profile fixed the HDR "problem".  We know why that is but I was under the impression that 709 was inherently narrow range.
